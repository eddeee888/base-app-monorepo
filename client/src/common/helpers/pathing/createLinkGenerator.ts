import { generatePath } from 'react-router';
import { UrlQuery } from './types';

interface LinkGenerator<P> {
  pattern: string;
  generate: (inputParams: P, urlQuery?: UrlQuery) => string;
}

const generateQueryString = (urlQuery?: UrlQuery): string => {
  if (!urlQuery) {
    return '';
  }

  let result = '?';

  (Object.keys(urlQuery) as (keyof UrlQuery)[]).forEach(queryKey => {
    result += `${queryKey}=${urlQuery[queryKey]}&`;
  });
  result = result.substring(0, result.length - 1);

  return result;
};

function createLinkGenerator<P>(pattern: string): LinkGenerator<P> {
  return {
    pattern,
    generate: (inputParams, urlQuery) => {
      return (
        generatePath(pattern, inputParams as any) +
        generateQueryString(urlQuery)
      );
    }
  };
}

export default createLinkGenerator;
