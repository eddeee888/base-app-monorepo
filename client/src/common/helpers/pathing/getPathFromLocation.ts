import { Paths } from './types';

const getPathFromLocation: (location: string) => Paths | null = location => {
  const strippedPath = stripUrlSpecialCharacterIfExists(
    stripUrlSpecialCharacterIfExists(
      stripUrlSpecialCharacterIfExists(location, '&'),
      '?'
    ),
    '#'
  );
  const path = '/' + strippedPath.split('/')[1];

  const resultArray = (Object.keys(Paths) as Array<keyof typeof Paths>).filter(
    key => Paths[key] === path
  );

  if (resultArray.length > 0) {
    return Paths[resultArray[0]];
  }
  return null;
};

const stripUrlSpecialCharacterIfExists: (
  path: string,
  characterToStrip: '?' | '#' | '&'
) => string = (path, characterToStrip) => {
  if (path.indexOf(characterToStrip) > -1) {
    return path.substring(0, path.indexOf(characterToStrip));
  }

  return path;
};

export default getPathFromLocation;
