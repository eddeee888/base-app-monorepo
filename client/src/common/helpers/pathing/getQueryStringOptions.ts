import {
  GetQueryStringOptionsFn,
  QueryStringKeys,
  QueryStringOptions
} from './types';

const isQueryStringKey = (key: string): key is keyof typeof QueryStringKeys => {
  const found = Object.keys(QueryStringKeys).filter(qsKey => qsKey === key);
  return found.length > 0;
};

const getQueryStringOptions: GetQueryStringOptionsFn = queryString => {
  const result: QueryStringOptions = {};

  const stripped = queryString.replace('?', '');
  const optionsString = stripped.split('&');
  optionsString.forEach(optionPair => {
    const [key, value] = optionPair.split('=');
    if (isQueryStringKey(key)) {
      result[key] = value;
    }
  });

  return result;
};

export default getQueryStringOptions;
