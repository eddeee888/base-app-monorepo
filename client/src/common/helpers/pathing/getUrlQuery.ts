import { GetUrlQueryFn, UrlQuery, UrlQueryKeys } from './types';

const isQueryStringKey = (key: string): key is keyof typeof UrlQueryKeys => {
  const found = Object.keys(UrlQueryKeys).filter(qsKey => qsKey === key);
  return found.length > 0;
};

const getUrlQuery: GetUrlQueryFn = queryString => {
  const result: UrlQuery = {};

  const stripped = queryString.replace('?', '');
  const optionsString = stripped.split('&');
  optionsString.forEach(optionPair => {
    const [key, value] = optionPair.split('=');
    if (isQueryStringKey(key)) {
      result[key] = value;
    }
  });

  return result;
};

export default getUrlQuery;
