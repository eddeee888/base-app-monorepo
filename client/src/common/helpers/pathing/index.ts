export { default as linkgen } from './linkgen';
export { default as getPathFromLocation } from './getPathFromLocation';
export { default as getUrlQuery } from './getUrlQuery';
export * from './types';
