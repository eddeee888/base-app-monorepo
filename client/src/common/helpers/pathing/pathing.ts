export enum Paths {
  home = '/',
  login = '/login',
  signup = '/signup',
  logout = '/logout',
  dashboard = '/dashboard',
  createClass = '/create-a-class'
}

interface QueryOptions {
  redirect?: string;
}

interface LinkgenOptions {
  query?: QueryOptions;
}

type LinkgenFn = (path: Paths, options?: LinkgenOptions) => string;

const linkgen: LinkgenFn = (path, options) => {
  let urlParts = '';

  if (options) {
    if (options.query) {
      urlParts += '?';
      (Object.keys(options.query) as Array<keyof QueryOptions>).forEach(
        queryKey => {
          urlParts += `${queryKey}=${options.query![queryKey]}&`;
        }
      );
      urlParts = urlParts.substring(0, urlParts.length - 1);
    }
  }

  return path + urlParts;
};

const stripUrlSpecialCharacterIfExists: (
  path: string,
  characterToStrip: '?' | '#' | '&'
) => string = (path, characterToStrip) => {
  if (path.indexOf(characterToStrip) > -1) {
    return path.substring(0, path.indexOf(characterToStrip));
  }

  return path;
};

const getPathFromLocation: (location: string) => Paths | null = location => {
  const strippedPath = stripUrlSpecialCharacterIfExists(
    stripUrlSpecialCharacterIfExists(
      stripUrlSpecialCharacterIfExists(location, '&'),
      '?'
    ),
    '#'
  );
  const path = '/' + strippedPath.split('/')[1];

  const resultArray = Object.keys(Paths).filter(key => {
    const typedKey = key as keyof typeof Paths;
    return Paths[typedKey] === path;
  });

  if (resultArray.length > 0) {
    const resultIndex = resultArray[0] as keyof typeof Paths;
    return Paths[resultIndex];
  }
  return null;
};

export { linkgen, getPathFromLocation };
