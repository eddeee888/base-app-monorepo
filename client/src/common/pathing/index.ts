export { default as useUrlQuery } from './useUrlQuery';
export * from './types';
