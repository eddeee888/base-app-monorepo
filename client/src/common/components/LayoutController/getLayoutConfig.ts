import { getPathFromLocation, linkgen, Paths } from 'common/helpers/pathing';

interface LayoutConfig {
  header: { display: boolean };
  footer: { display: boolean };
}

type GetLayoutConfigFn = (pathname: string) => LayoutConfig;

const getLayoutConfig: GetLayoutConfigFn = pathname => {
  const path = getPathFromLocation(pathname);

  let header = { display: true };
  let footer = { display: true };

  switch (path) {
    case linkgen(Paths.signup):
      header = {
        display: true
      };
      footer = {
        display: true
      };
      break;
    case linkgen(Paths.login):
      header = {
        display: true
      };
      footer = {
        display: true
      };
      break;
    default:
      break;
  }

  return {
    header,
    footer
  };
};

export default getLayoutConfig;
