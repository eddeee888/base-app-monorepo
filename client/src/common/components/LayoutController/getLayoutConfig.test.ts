import { Paths } from 'common/helpers/pathing';
import getLayoutConfig from './getLayoutConfig';

describe('getLayoutConfig()', () => {
  (Object as any).values(Paths).forEach((path: string) => {
    it(`should match snapshot for '${path}' path`, () => {
      const config = getLayoutConfig(path);
      expect(config).toMatchSnapshot();
    });
  });

  it('should match snapshot if invalid path', () => {
    const config = getLayoutConfig('/Thisishighlyinvalidpath');
    expect(config).toMatchSnapshot();
  });
});
