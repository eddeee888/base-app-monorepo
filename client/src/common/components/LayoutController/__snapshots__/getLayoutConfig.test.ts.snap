// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getLayoutConfig() should match snapshot for '/' path 1`] = `
Object {
  "footer": Object {
    "display": true,
  },
  "header": Object {
    "display": true,
  },
}
`;

exports[`getLayoutConfig() should match snapshot for '/dashboard' path 1`] = `
Object {
  "footer": Object {
    "display": true,
  },
  "header": Object {
    "display": true,
  },
}
`;

exports[`getLayoutConfig() should match snapshot for '/login' path 1`] = `
Object {
  "footer": Object {
    "display": true,
  },
  "header": Object {
    "display": true,
  },
}
`;

exports[`getLayoutConfig() should match snapshot for '/logout' path 1`] = `
Object {
  "footer": Object {
    "display": true,
  },
  "header": Object {
    "display": true,
  },
}
`;

exports[`getLayoutConfig() should match snapshot for '/signup' path 1`] = `
Object {
  "footer": Object {
    "display": true,
  },
  "header": Object {
    "display": true,
  },
}
`;

exports[`getLayoutConfig() should match snapshot if invalid path 1`] = `
Object {
  "footer": Object {
    "display": true,
  },
  "header": Object {
    "display": true,
  },
}
`;
